### PR TITLE
improve error info for arrow function at right of binary expression

### LINKED ISSUE #25898: Improve errors for potential arrow functions
There are certain locations where arrow functions don't parse

```ts
var x = window.foo || () => {}
```

In these instances, it would be nice if users could get a better error message as to avoid issues like #25897.

Ideally this should not be an invasive change.

* The change should be limited in scope.
* The parser should continue erroring in those locations.
* The change must not introduce change how existing valid code is parsed.

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #25898

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -6015,6 +6026,14 @@ namespace ts {
             return nextToken() === SyntaxKind.SlashToken;
         }
 
+        function isStartOfSimpleArrowFunction () {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Nit: not space before parentheses

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -6015,6 +6026,14 @@ namespace ts {
             return nextToken() === SyntaxKind.SlashToken;
         }
 
+        function isStartOfSimpleArrowFunction () {
+            if (token() === SyntaxKind.AsyncKeyword) {
+                nextToken();
+                return isIdentifier() && nextToken() === SyntaxKind.EqualsGreaterThanToken || token() === SyntaxKind.EqualsGreaterThanToken;
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Can you leave a comment on the intent here? Is this supposed to cover `async => 100`?

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -4125,6 +4125,10 @@
         "category": "Error",
         "code": 17017
     },
+    "Invalid arrow-function arguments, parentheses around the arrow-function may help.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]:
```
Arrow functions are not syntactically valid here. Consider wrapping the function in parentheses.
```

File: tests/baselines/reference/incorrectArrowFunction.errors.txt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+!!! error TS2304: Cannot find name 'async'.
+                     ~~
+!!! error TS1005: ';' expected.
+    true || async a => 1;
+           ~~~~~~~~~~~~~
+!!! error TS17018: Invalid arrow-function arguments, parentheses around the arrow-function may help.
+                        ~
+!!! error TS1109: Expression expected.
+    true || async => 1;
+           ~~~~~~~~~~~
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/incorrectArrowFunction.errors.txt]: You really don't want the error span to include whitespace.

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -3791,7 +3791,18 @@ namespace ts {
...
+                        const maybeArrow = lookAhead(tryParseParenthesizedArrowFunctionExpression);
+                        if (maybeArrow) {
+                            parseErrorAtRange(maybeArrow, Diagnostics.Invalid_arrow_function_arguments_parentheses_around_the_arrow_function_may_help);
+                        }
+                    }
+                    else if (lookAhead(isStartOfSimpleArrowFunction)) {
+                        const maybeArrow = tryParseAsyncSimpleArrowFunctionExpression() || lookAhead(() => parseSimpleArrowFunctionExpression(<Identifier>parseBinaryExpressionOrHigher(/*precedence*/ 0)));
+                        parseErrorAtRange(maybeArrow, Diagnostics.Invalid_arrow_function_arguments_parentheses_around_the_arrow_function_may_help);
+                    }
+                    leftOperand = makeBinaryExpression(leftOperand, op, parseBinaryExpressionOrHigher(newPrecedence));
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: You're currently ignoring it if it's present, but I think that you actually want to gracefully parse the arrow functions you've been able to find above.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -3791,7 +3791,18 @@ namespace ts {
                     }
                 }
                 else {
-                    leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                    const op = parseTokenNode<Token<BinaryOperator>>();
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Leave a comment like

```
// Much of the time, users will write code along the lines of
//
//   let x = foo || () => { /*...*/ }
//
// However, arrow functions aren't valid in those positions in the ECMAScript grammar.
// Despite that, we'll try to parse it out anyway and give a decent error message.
```

**Comment by RyanCavanaugh** [comment]
Many unaddressed comments - please open a fresh PR if desired. Thanks!

**Pull request closed without merging**